quickwit-storage/src/cache/in_ram_slice_cache.rs

```diff
@@ -0,0 +1,213 @@
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+
+use bytes::Bytes;
+use lru::LruCache;
+use tracing::{error, warn};
+
+#[derive(Hash, Debug, Clone, PartialEq, Eq)]
+struct SliceAddress {
+    pub path: PathBuf,
+    pub byte_range: Range<usize>,
+}
+
+struct NeedMutSliceCache {
+    lru_cache: LruCache<SliceAddress, Bytes>,
+    num_bytes: usize,
+    capacity_in_bytes_opt: Option<usize>,
+}
+
+impl NeedMutSliceCache {
+    fn with_capacity_in_bytes(capacity_in_bytes: Option<usize>) -> Self {
+        NeedMutSliceCache {
+            // The limit will be decided by the amount of memory in the cache,
+            // not the number of items in the cache.
+            // Enforcing this limit is done in the `NeedMutCache` impl.
+            lru_cache: LruCache::unbounded(),
+            num_bytes: 0,
+            capacity_in_bytes_opt: capacity_in_bytes,
+        }
+    }
+
+    fn get(&mut self, cache_key: &SliceAddress) -> Option<Bytes> {
+        self.lru_cache.get(cache_key).cloned()
+    }
+
+    fn greater_than_capacity(&self, len: usize) -> bool {
+        if let Some(capacity_in_bytes) = self.capacity_in_bytes_opt {
+            len > capacity_in_bytes
+        } else {
+            false
+        }
+    }
+
+    /// Attempt to put the given amount of data in the cache.
+    /// This may fail silently if the owned_bytes slice is larger than the cache
+    /// capacity.
+    fn put(&mut self, slice_addr: SliceAddress, bytes: Bytes) {
+        if self.greater_than_capacity(bytes.len()) {
+            // The value does not fit in the cache. We simply don't store it.
+            warn!(
+                capacity_in_bytes = ?self.capacity_in_bytes_opt,
+                len = bytes.len(),
+                "Downloaded a byte slice larger than the cache capacity."
+            );
+            return;
+        }
+        if let Some(previous_data) = self.lru_cache.pop(&slice_addr) {
+            self.num_bytes -= previous_data.len();
+        }
+        while self.greater_than_capacity(self.num_bytes + bytes.len()) {
+            if let Some((_, bytes)) = self.lru_cache.pop_lru() {
+                self.num_bytes -= bytes.len();
+            } else {
+                error!("Logical error. Even after removing all of the items in the cache the capacity is insufficient. This case is guarded against and should never happen.");
+                return;
+            }
+        }
+        self.num_bytes += bytes.len();
+        self.lru_cache.put(slice_addr, bytes);
+    }
+}
+
+/// A simple in-resident memory slice cache.
+pub struct SliceCache {
+    inner: Mutex<NeedMutSliceCache>,
+}
+
+impl SliceCache {
+    /// Creates an slice cache with the given capacity.
+    ///
+    /// The SliceCache is guaranteed to not store more than `capcity_in_bytes`.
+    /// If None, SliceCache acts as if the capacity is unlimited.
```

Comment:
nitpick:
I think this doc comment should go on `NeedMutSliceCache`.  we already have `with_capacity_in_bytes` and `with_infinite_capacity` methods that explain.  My point is because `capcity_in_bytes` here is not an Option

Reply:
yes you are right